chacha8rand.md

```diff
@@ -0,0 +1,245 @@
+
+```
+stream = ChaCha8(key = input, nonce = {0}, len = 1024)
+for (i = 0; i < 16; i++) // for each block
+	stream[i*64 + 0*4 : i*64 + 0*4 + 4] -= 0x61707865
+	stream[i*64 + 1*4 : i*64 + 1*4 + 4] -= 0x3320646e
+	stream[i*64 + 2*4 : i*64 + 2*4 + 4] -= 0x79622d32
+	stream[i*64 + 3*4 : i*64 + 3*4 + 4] -= 0x6b206574
+	stream[i*64 + 12*4 : i*64 + 12*4 + 4] -= i
+output = ""
+for (b = 0; b < 16; n += 4) // for each sequence of four blocks
+	for (i = 0; i < 64; i += 4) // for each four bytes in a block
+		output += stream[0*64 + i : 0*64 + i + 4]
+		output += stream[1*64 + i : 1*64 + i + 4]
+		output += stream[2*64 + i : 2*64 + i + 4]
+		output += stream[3*64 + i : 3*64 + i + 4]
+	stream = stream[256:]
+next_input = output[1024-32:]
+output = output[:1024-32]
+```
+
+Note that the subtractions are equivalent to skipping the final state addition for the constant, nonce, and counter terms.
+
+## Design rationale
+
+**Why the permutation?** On platforms with SIMD registers that can hold 128 bits, ChaCha can be parallelized four-ways (that is, producing four blocks at a time) by packing the first uint32 of state for each block in the first SIMD register, and so on. Defining the output to be permuted like in ChaCha8Rand allows producing it without having to deinterlace it, by simply concatenating the packed SIMD registers.
+
+**Why sixteen blocks at a time?** All amd64 processors have 128-bit vector registries, so we can target them without CPU feature detection. Other implementers might find it preferable to pay the complexity and buffer size cost to parallelize eight or sixteen blocks, with 256-bit and 512-bit vector instructions respectively. We produce sixteen blocks before rekeying but interlace four to optimize for a smaller buffer (of four block at a time) but still future-proof (by allowing up to sixteen blocks to be generated at a time with 512-bit vector instructions) and improve throughput.
+
+**Key erasure vs state compression.** Traditional key erasure involves immediately overwriting the key, and deleting buffered output bytes as soon as they are produced. An alternative strategy is deferring the overwriting until the next iteration starts, in which case the state can be serialized to just the last iteration input (from which the current buffer and the next iteration key can be regenerated) and a counter.
```

Comment:
This, combined with the 33 byte length specified above, indicates that this is a 1-byte counter. But nowhere is it specified what this is counting.

My guess is that this is meant to be the ChaCha8 counter (which is guaranteed to fit into 1 byte), rather than a count of how many of the 992 bytes per rekey have been output (which would require a 2-byte counter). But the unspecified implication is that when randomness is sampled that only consumes part of a block, the remainder of the block is discarded. That in turn means that the sample output is only going to match for continuous reads (or reads of a multiple of 992 bytes).

Reply:
Opened #47.